### PR TITLE
SamlAttributeConverter config option for pac4j

### PR DESCRIPTION
**Pull request by fbusselgln**
Hey there!

[This pac4j change](https://github.com/pac4j/pac4j/commit/ec734d7759080eb87ccccbfac1a07dde924bf4bc) changed the behaviour of handling XML complex type attributes in pac4j-saml. The commit introduced a new configuration option, which hasn't made it's way into the CAS configuration, yet. However we need the non-default behaviour for our project.

This pull request adds a new configuration option for the Pac4j feature to CAS without changing the default behaviour.

I am not sure whether you require extra unit tests for this, as the configuration is covered by plenty of test cases already?!

### CODE REVIEW

**Comment by mmoayyed** [comment]
Thank you for the patch.

Yes we require unit tests for ALL changes, and if you are up to it, it would be great to add functional tests as well.

File: api/cas-server-core-api-configuration-model/src/main/java/org/apereo/cas/configuration/model/support/pac4j/saml/Pac4jSamlClientProperties.java [hunk 1]
```diff
@@ -289,6 +289,15 @@ public class Pac4jSamlClientProperties extends Pac4jBaseClientProperties impleme
...
     private String messageStoreFactory = "org.pac4j.saml.store.EmptyStoreFactory";
 
+    /**
+     * Controls the way SAML attributes are converted from the SAML authentication response into pac4j attributes.
+     * <ul>
+     *     <li>{@code org.pac4j.saml.profile.converter.SimpleSAML2AttributeConverter}: Values of complex types are serialized into a single attribute. (default)</li>
+     *     <li>{@code org.pac4j.saml.profile.converter.ComplexTypeSAML2AttributeConverter}: Values of complex types are unboxed into several attributes; one per child element.</li>
+     * </ul>
+     */
+    private String samlAttributeConverter = "org.pac4j.saml.profile.converter.SimpleSAML2AttributeConverter";
```
Comment (mmoayyed) [inline api/cas-server-core-api-configuration-model/src/main/java/org/apereo/cas/configuration/model/support/pac4j/saml/Pac4jSamlClientProperties.java:299]: Does this really need to be the default? What happens if you remove the value here?

File: api/cas-server-core-api-configuration-model/src/main/java/org/apereo/cas/configuration/model/support/pac4j/saml/Pac4jSamlClientProperties.java [hunk 2]
```diff
@@ -289,6 +289,15 @@ public class Pac4jSamlClientProperties extends Pac4jBaseClientProperties impleme
      */
     private String messageStoreFactory = "org.pac4j.saml.store.EmptyStoreFactory";
 
+    /**
+     * Controls the way SAML attributes are converted from the SAML authentication response into pac4j attributes.
+     * <ul>
+     *     <li>{@code org.pac4j.saml.profile.converter.SimpleSAML2AttributeConverter}: Values of complex types are serialized into a single attribute. (default)</li>
```
Comment (mmoayyed) [inline api/cas-server-core-api-configuration-model/src/main/java/org/apereo/cas/configuration/model/support/pac4j/saml/Pac4jSamlClientProperties.java:295]: You should either turn these into @link and make sure the javadoc is valid, or simply remove the impls and refer to the actual Converter interface. This is not maintainable in this form specially if the impls move way or get renamed, etc, we will not be available to catch them. A link to pac4j documentation page would also be appropriate.

File: support/cas-server-support-pac4j-core-clients/src/main/java/org/apereo/cas/support/pac4j/authentication/clients/BaseDelegatedClientFactory.java [hunk 3]
```diff
@@ -568,6 +569,14 @@ protected Collection<IndirectClient> buildSaml2IdentityProviders(final CasConfig
                 cfg.setProviderName(saml.getProviderName());
                 cfg.setNameIdPolicyAllowCreate(saml.getNameIdPolicyAllowCreate().toBoolean());
 
+                if (!saml.getSamlAttributeConverter().isEmpty() && saml.getSamlAttributeConverter().contains(".")) {
```
Comment (mmoayyed) [inline support/cas-server-support-pac4j-core-clients/src/main/java/org/apereo/cas/support/pac4j/authentication/clients/BaseDelegatedClientFactory.java:572]: Is the second check necessary?

**Review by mmoayyed — CHANGES_REQUESTED** [review]
Looks great to me. See comments please.

File: api/cas-server-core-api-configuration-model/src/main/java/org/apereo/cas/configuration/model/support/pac4j/saml/Pac4jSamlClientProperties.java (hunk 1 above)
Comment (fbusselgln) [reply api/cas-server-core-api-configuration-model/src/main/java/org/apereo/cas/configuration/model/support/pac4j/saml/Pac4jSamlClientProperties.java:299]: You are right, the default pac4j configuration will use the SimpleSAML2AttributeConverter anyway, so defining a specific class via the configuration will do the trick if you want to enforce non-default behaviour.

**Comment by fbusselgln** [comment]
Fixes applied and unit test added. I hope this looks ok to you?

File: support/cas-server-support-pac4j-core-clients/src/test/java/org/apereo/cas/support/pac4j/authentication/clients/DelegatedClientFactoryTests.java [hunk 4]
```diff
@@ -0,0 +1,73 @@
...
+        CasCoreUtilConfiguration.class,
+        CasCoreHttpConfiguration.class
+}, properties = {
+        "cas.authn.pac4j.saml[0].clientName=foo_client",
+        "cas.authn.pac4j.saml[0].identityProviderMetadataPath=file:/idp-metadata.xml",
+        "cas.authn.pac4j.saml[0].samlAttributeConverter=org.apereo.cas.support.pac4j.authentication.clients.DelegatedClientFactoryTests.FooConverter",
+})
+@EnableConfigurationProperties(CasConfigurationProperties.class)
+@Tag("SAML")
+public class DelegatedClientFactoryTests {
```
Comment (mmoayyed) [inline support/cas-server-support-pac4j-core-clients/src/test/java/org/apereo/cas/support/pac4j/authentication/clients/DelegatedClientFactoryTests.java:40]: Please remove this, and move tests to `DefaultDelegatedClientFactoryTests`.

Comment (fbusselgln) [reply support/cas-server-support-pac4j-core-clients/src/test/java/org/apereo/cas/support/pac4j/authentication/clients/DelegatedClientFactoryTests.java]: Done

**Comment by fbusselgln** [comment]
Something else: Do you think a backport for this to the 6.6.0 branch would be possible once this is merged?

**Comment by mmoayyed** [comment]
Yes backport is possible for sure. You're welcome to do it as a separate PR, or I could cherry pick for you on your behalf once this is merged.

**Comment by mmoayyed** [comment]
@apereocas-bot runci

**Comment by mmoayyed** [comment]
All looks good, thank you. I have triggered CI to do its thing and should be able to merge quickly afterwards. Thanks much!

**Comment by fbusselgln** [comment]
While you're on it, maybe you could indeed cherry pick the squashed commit as backport for 6.6.x? I guess there are plenty of potential mistakes I could make with that PR as well... ;)

**Comment by mmoayyed** [comment]
@apereocas-bot runci

**Comment by mmoayyed** [comment]
No problem. I will cherry pick.

**Pull request merged**